internal/controllers/reference/reference.go

```diff
@@ -0,0 +1,150 @@
+// secrets that are really referred by referrer is passed in parameter referredSecretNames.
+// if the secret is not actually referrenced by any object after deleting outdated reference records,
+// the secret will be removed from object cache in dataplaneClient.
+func removeOutdatedReferencesToSecret(
+	ctx context.Context,
+	indexers CacheIndexers, c client.Client, dataplaneClient *dataplane.KongClient,
+	referrer client.Object, referredSecretNames []types.NamespacedName,
+) error {
+	referredSecretNameMap := make(map[types.NamespacedName]bool, len(referredSecretNames))
+	for _, nsName := range referredSecretNames {
+		referredSecretNameMap[nsName] = true
+	}
+
+	referents, err := indexers.ListReferredObjects(referrer)
+	if err != nil {
+		return err
+	}
+	for _, obj := range referents {
+		gvk := obj.GetObjectKind().GroupVersionKind()
+		// delete the reference record if the secret is not referred by the service.
+		if gvk.Group == corev1.GroupName && gvk.Version == VersionV1 && gvk.Kind == KindSecret {
+			namespacedName := types.NamespacedName{
+				Namespace: obj.GetNamespace(),
+				Name:      obj.GetName(),
+			}
+
+			// if the secret is still referenced, no operations are taken so continue here.
+			if referredSecretNameMap[namespacedName] {
+				continue
+			}
+
+			if err := indexers.DeleteObjectReference(referrer, obj); err != nil {
+				return err
+			}
+			// remove the secret in object cache if it is not referred and does not have label "konghq.com/ca-cert:true".
+			// Do this check and delete when the reference count may be reduced by 1.
+
+			// retrieve the secret in k8s and check it has the label.
+			secret := &corev1.Secret{}
+			getErr := c.Get(ctx, namespacedName, secret)
+			// if the secret exists in k8s and has the label, we should not delete it in object cache.
+			if getErr == nil {
+				if secret.Labels != nil && secret.Labels[CACertLabelKey] == "true" {
+					continue
+				}
+			} else {
+				// if the secret does not exist in k8s, we ignore the error and continue the check and delete operation.
+				// for other errors, we return the error and stop the operation.
+				if !k8serrors.IsNotFound(getErr) {
+					return err
+				}
+			}
+
+			if err := indexers.DeleteObjectIfNotReferred(obj, dataplaneClient); err != nil {
+				return err
+			}
+
+		}
+	}
+	return nil
+}
+
+// DeleteReferencesByReferrer deletes all reference records with specified referrer
+// in reference cache.
+// If the affected secret is not referred by any other objects, it deletes the secret in object cache.
+func DeleteReferencesByReferrer(indexers CacheIndexers, dataplaneClient *dataplane.KongClient, referrer client.Object) error {
+	referents, err := indexers.ListReferredObjects(referrer)
+	if err != nil {
+		return err
+	}
+
+	// delete(gc) the reference record between referrer and referent.
+	for _, referent := range referents {
+		err := indexers.DeleteObjectReference(referrer, referent)
```

Comment:
What happens if `indexers.DeleteObjectReference` returns an error. Will the `referents` be deleted next time?